mitmproxy/addonmanager.py

```diff
@@ -273,6 +274,38 @@ def invoke_addon(self, addon, event: hooks.Hook):
                         f"Addon handler {event.name} ({a}) not callable"
                     )
 
+    async def async_invoke_addon(self, addon, event: hooks.Hook):
+        """
+            Asynchronously invoke an event on an addon and all its children.
+        """
+        for addon, func in self._iter_hooks(addon, event):
+            res = func(*event.args())
+            # Support both async and sync hook functions
+            if isawaitable(res):
+                await res
+
+    def invoke_addon(self, addon, event: hooks.Hook):
+        """
+            Invoke an event on an addon and all its children.
+        """
+        for addon, func in self._iter_hooks(addon, event):
+            res = func(*event.args())
+            if isawaitable(res):
+                raise exceptions.AddonManagerError(
+                    f"Async handler {event.name} ({addon}) called from sync context"
```

Comment:
FWWI I think this is quite a substantial limitation. The long-term plan needs to be that all callers (including commands) should be async-capable. That of course gets unwieldy for a single PR, so let's keep this error here for now. Just FYI.

Reply:
Yes, it is a big limitation. In particular, I noticed that `run.script` will be unable to call async methods, which sucks for replayability :(.

It might be nice to have async commands for 8.0, although that's trickier because callers would then need to be async-aware themselves.